Implementing a service for interacting with the api

Danila-Peshev opened the pull request:
Task 3

Danila-Peshev committed:
added a service for interacting with the API with methods according to the requirements of the task

Danila-Peshev committed:
A global LanguageContext has been created for working with the API;
New types added;
Queries in MovieService have been updated;
Added an EXAMPLE of using the service functions on the home page;
 Existing folders have been renamed and new folders have been added.

Danila-Peshev committed:
unfinished commit
add simple movie type for show view;
add page for viewing information about the movie.

Danila-Peshev committed:
Comleted commit:
All types moved to directory;
Has been LanguageProvider implemented for globality;
Test data paced in .env.example;
Improved methods in MovieService;
Everythinc do not relate this task has been deleted.

Danila-Peshev committed:
Improve service:
Add timeout(15s);
Handling errors MovieService.

Isidora25 commented on src/components/switch-language/LanguageContext.tsx:
```diff
@@ -0,0 +1,31 @@
+import { Language } from "../../types/Language";
+
+type LanguageContextType = {
+  activeLanguage: Language;
```
```suggestion
  language: Language;
```

Isidora25 commented on src/services/MovieService.ts:
```diff
@@ -0,0 +1,148 @@
+    primary_release_year: releaseYear,
+  });
+
+  const simpleMovies: SimpleMovie[] = data.results.map((movie: Movie) => ({
+    id: movie.id,
+    backdrop_path: movie.backdrop_path,
+    genre_ids: movie.genre_ids,
+    overview: movie.overview,
+    popularity: movie.popularity,
+    poster_path: movie.poster_path,
+    release_date: movie.release_date,
+    title: movie.title,
+    vote_average: movie.vote_average,
+    vote_count: movie.vote_count,
+  }));
```
не оч поняла задумку. но если уж тебе очень надо мапать, переписывай сразу в camelCase

Danila-Peshev replied:
Привожу все полученные фильмы с API к SimpleMovie, чтобы не было лишней информации

Isidora25 commented on src/services/MovieService.ts:
```diff
@@ -0,0 +1,148 @@
+  }));
+
+  return {
+    page: data.page,
+    results: simpleMovies,
+    total_pages: data.total_pages,
+    total_result: data.total_results,
```
то же замечание про camelCase

Danila-Peshev replied:
Я старался сделать названия такими же, как у TMBD. Хорошо, переименую

Isidora25 commented on src/services/MovieService.ts:
```diff
@@ -0,0 +1,148 @@
+  activeLanguage?: Language,
+  page?: number
+} = {}): Promise<MoviesResponse> {
+  if (ids.length === 0) {
+    return {
+      page,
+      results: [],
+      total_pages: 0,
+      total_result: 0,
+    };
+  }
```
вообще странно вызывать функцию, которая отдает что-то по id не передавая туда никаких id 
если id нет, то функцию вообще не стоит вызывать, а параметр ids должен быть обязательным

Isidora25 commented on src/services/MovieService.ts:
```diff
@@ -0,0 +1,148 @@
+
+  const results = await Promise.all(ids.map(id => getMovieById({ id, activeLanguage })));
+
+  const total_result = results.length;
+  const total_pages = Math.ceil(total_result / 20);
+
+  const startIndex = (page - 1) * 20;
+  const endIndex = startIndex + 20;
+
+  const paginatedResults = results.slice(startIndex, endIndex);
```
20 -> magic number, непонятно, что это такое. такие числовые значения обычно выносят в переменные с говорящим именем.

Isidora25 commented on src/types/movie-api-types/OnlyMovie.ts:
```diff
@@ -0,0 +1,48 @@
+import { Genre } from "./Genre"
+
+export interface OnlyMovie {
+  adult: boolean
+  backdrop_path: string
+  belongs_to_collection: any
+  budget: number
+  genres: Genre[]
+  homepage: string
+  id: number
+  imdb_id: string
+  origin_country: string[]
+  original_language: string
+  original_title: string
+  overview: string
+  popularity: number
+  poster_path: string
+  production_companies: ProductionCompany[]
+  production_countries: ProductionCountry[]
+  release_date: string
+  revenue: number
+  runtime: number
+  spoken_languages: SpokenLanguage[]
+  status: string
+  tagline: string
+  title: string
+  video: boolean
+  vote_average: number
+  vote_count: number
```
тебе не нужно описывать ВЕСЬ объект, который от апи возвращается 
опиши, только то, что нужно

Danila-Peshev committed:
Commit:
Change language name;
delete large type OnlyMovies;
add variable maxMoviesOnPage;
Rename fields type SimpleMovie and MovieResponse as camelCase;

Isidora25 commented on src/services/MovieService.ts:
```diff
@@ -38,19 +37,19 @@ async function fetchFromApi(endpoint: string, params: object): Promise<any> {
-export async function getAllGenres(activeLanguage: string): Promise<Genre[]> {
-  const data = await fetchFromApi('/genre/movie/list', { language: activeLanguage });
+export async function getAllGenres(language: string): Promise<Genre[]> {
+  const data = await fetchFromApi('/genre/movie/list', { language: language });
```
```suggestion
  const data = await fetchFromApi('/genre/movie/list', { language });
```

Isidora25 commented on src/components/switch-language/LanguageSwitcher.tsx:
```diff
@@ -4,23 +4,23 @@ import SwitchLanguageButton from "./SwitchLanguageButton";
-  const { activeLanguage, changeLanguage } = useLanguage();
+  const { language, changeLanguage } = useLanguage();
   const { i18n } = useTranslation();
 
   useEffect(() => {
-    i18n.changeLanguage(activeLanguage);
-  }, [activeLanguage, i18n]);
+    i18n.changeLanguage(language);
+  }, [language, i18n]);
```
это тоже все двинь в контекст

Isidora25 commented on src/services/MovieService.ts:
```diff
@@ -69,80 +68,73 @@ export async function getMoviesByFilters({
-    vote_average: movie.vote_average,
-    vote_count: movie.vote_count,
+    voteAverage: movie.vote_average,
+    voteCount: movie.vote_count,
```
тут можно короче с деструктуризацией - 
```
const simpleMovies: SimpleMovie[] = data.results.map(({id,backdrop_path,genre_ids, overview  }: Movie) => ({
     id, 
     backdropPath: movie.backdrop_path,
    genre_ids: movie.genre_ids,
    overview
```

Isidora25 commented on src/services/MovieService.ts:
```diff
@@ -69,80 +68,73 @@ export async function getMoviesByFilters({
 
-  const results = await Promise.all(ids.map(id => getMovieById({ id, activeLanguage })));
+  const results = await Promise.all(ids.map(id => getMovieById({ id, language })));
+  const maxMoviesOnPage = 20;
```
выноси неизменяемые константы за функцию и используй UPPER_SNAKE_CASE

Isidora25 commented on src/services/MovieService.ts:
```diff
@@ -69,80 +68,73 @@ export async function getMoviesByFilters({
-    backdrop_path: data.backdrop_path,
-    genre_ids: data.genres.map((genre) => genre.id),
+    backdropPath: data.backdrop_path,
+    genreIds: data.genres.map((genre: { id: number; }) => genre.id),
```
```suggestion
    genreIds: data.genres.map(( { id }) =>id),
```

Danila-Peshev committed:
Commit:
Move useEffect to LanguageProvider;
Took out MAX_MOVIES_ON_PAGE;
Short the destructurization;

Isidora25 commented on src/services/MovieService.ts:
```diff
@@ -66,18 +68,31 @@ export async function getMoviesByFilters({
+    }: Movie) => ({
+      id,
+      backdrop_path,
+      genre_ids,
```
а куда пропал camelCase?)

Danila-Peshev committed:
Rename field as camelCase

reshetov commented on .env.example:
```diff
@@ -0,0 +1,2 @@
+REACT_APP_ACCESS_TOKEN_AUTH=eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiI5NjIwNmU1MDYxMjdhY2FlMjU4NTQ2MDA3N2VhMTRmOSIsIm5iZiI6MTcyMjEwMzUyOS4xMTQwNSwic3ViIjoiNjY5ZThiMzY2NzkzMDMxN2RmZjM3NmU0Iiwic2NvcGVzIjpbImFwaV9yZWFkIl0sInZlcnNpb24iOjF9.ST3hPNsOLRtdIBiRglvH_xQB381JK4KGx1itavpTyj0
```
Если это боевой токен убери его от греха. Использую заглушки вида `xxxxxxxx`
